segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+     * </li>
+     * <li>
+     * What is the defrag algorithm
+     * <pre>
+     * While(segment.hasConcatableChunks()){
+     *     Set<List<Chunk>> s = FindConsecutiveConcatableChunks();
+     *     For (List<chunk> list : s){
+     *        ConcatChunks (list);
+     *     }
+     * }
+     * </pre>
+     * </li>
+     * </ul>
+     *
+     * @param txn             Active {@link MetadataTransaction}.
+     * @param segmentMetadata {@link SegmentMetadata} for the segment to defrag.
+     * @param startChunkName  Name of the first chunk to start defragmentation.
+     * @param lastChunkName   Name of the last chunk before which to stop defragmentation. (last chunk is not concatenated).
+     * @throws Exception In case of any errors.
+     */
+    private void defrag(MetadataTransaction txn, SegmentMetadata segmentMetadata, String startChunkName, String lastChunkName) throws Exception {
+        // The algorithm is actually very simple.
+        // It tries to concat all small chunks using appends first.
+        // Then it tries to concat remaining chunks using native concat if available.
+        // To implement it using single loop we toggle between concat with append and native concat modes. (Instead of two passes.)
+        boolean useAppend = true;
+        String targetChunkName = startChunkName;
+
+        // Iterate through chunk list
+        while (null != targetChunkName && !targetChunkName.equals(lastChunkName)) {
+            ChunkMetadata target = (ChunkMetadata) txn.get(targetChunkName);
+
+            ArrayList<ChunkInfo> chunksToConcat = new ArrayList<>();
+            long targetSizeAfterConcat = target.getLength();
+
+            // Add target to the list of chunks
+            chunksToConcat.add(new ChunkInfo(targetSizeAfterConcat, targetChunkName));
+
+            String nextChunkName = target.getNextChunk();
+            ChunkMetadata next = null;
+
+            // Gather list of chunks that can be appended together.
+            while (null != nextChunkName) {
+                next = (ChunkMetadata) txn.get(nextChunkName);
+
+                if (useAppend && config.getMinSizeLimitForNativeConcat() < next.getLength()) {
+                    break;
+                }
+
+                if (targetSizeAfterConcat + next.getLength() > segmentMetadata.getMaxRollinglength() || next.getLength() > config.getMaxSizeLimitForNativeConcat()) {
+                    break;
+                }
+
+                chunksToConcat.add(new ChunkInfo(next.getLength(), nextChunkName));
+                targetSizeAfterConcat += next.getLength();
+
+                nextChunkName = next.getNextChunk();
+            }
+            // Note - After above while loop is exited nextChunkName points to chunk next to last one to be concat.
+            // Which means target should now point to it as next after concat is complete.
+
+            // If there are chunks that can be appended together then concat them.
```

Comment:
`If there are chunks that can be appended together, then concat them.`

Reply:
?